.github/workflows/mainPipeline.yml

```diff
@@ -0,0 +1,147 @@
+name: Main pipeline
+on:
+  push:
+    branches:
+      - main
+  pull_request:
+    branches: [ main ]
+  workflow_dispatch:
+    inputs:
+      version_override:
+        description: 'Version Number'     
+        required: true
+        default: '0.0.0'
+
+jobs:
+  draftRelease:
+    name: Draft release
+    runs-on: ubuntu-latest
+    steps:
+      # Drafts your next Release notes as Pull Requests are merged into "master"
+      - uses: release-drafter/release-drafter@v5
+        if: github.event_name == 'push'
+        env:
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+
+  buildTest:
+    name: Build and test
+    runs-on: windows-latest 
+    needs: draftRelease
+    outputs:
+      buildNumber: ${{ steps.ver.outputs.BUILD_NUMBER }}
+    steps:
+      - uses: actions/checkout@v2
+        with:
+          fetch-depth: 0  # Shallow clones should be disabled for a better relevancy of analysis
+
+      - name: GetVersion
+        id: ver
+        shell: pwsh
+        run: |
+              if ("${{github.event.inputs.version_override}}")
+              {
+                Write-Host "Using manually entered override version"
+                $version = "${{ github.event.inputs.version_override}}"
+              }
+              else
+              {
+                $header = @{Authorization = 'Bearer ${{ secrets.GITHUB_TOKEN }}' }
+                $endPoint = "https://api.github.com/repos/${{github.repository}}/releases"
+                $info = Invoke-RestMethod -Uri $endPoint -Headers $header
+                $version = $info.name[0].Trim("v")
+              }
+              
+              Write-Host "::set-output name=BUILD_NUMBER::$($version)"
+          
+      - name: Set up JDK 11
+        uses: actions/setup-java@v1
+        with:
+          java-version: 1.11
+      
+      - name: Setup .NET Core
+        uses: actions/setup-dotnet@v1
+        with:
+          dotnet-version: 3.1.101
+
+      - name: Install dependencies
+        run: dotnet restore Selenium.Axe/Selenium.Axe.sln
+        
+      - name: Update Firefox
+        run: choco upgrade firefox
+        
+      - name: Update Chrome
+        run: choco upgrade googlechrome
+        
+      - name: Cache SonarCloud packages
+        uses: actions/cache@v1
+        with:
+          path: ~\sonar\cache
+          key: ${{ runner.os }}-sonar
+          restore-keys: ${{ runner.os }}-sonar
+          
+      - name: Cache SonarCloud scanner
+        id: cache-sonar-scanner
+        uses: actions/cache@v1
+        with:
+          path: .\.sonar\scanner
+          key: ${{ runner.os }}-sonar-scanner
+          restore-keys: ${{ runner.os }}-sonar-scanner
```

Comment:
This is going to lock to a potentially-outdated version of the scanner basically indefinitely. Perhaps the key here should include a version number for the scanner tool?

Reply:
Not sure how we would get stuck on an outdated version, we are aren't hosting our own agents.

Reply:
Cached directories can be shared between different builds, even using GitHub-provided agents. Because the `Install SonarCloud scanner` step uses `if: steps.cache-sonar-scanner.outputs.cache-hit != 'true'`, it's going to be skipped if there exists a previously-cached `.\.sonar\scanner` directory with key `${{ runner.os }}-sonar-scanner`. That means that the first time a build runs, it'll cache the scanner folder with key `Windows-sonar-scanner`, and since that key name isn't based on the scanner version, whatever version was cached will keep being re-used in subsequent builds (until the cache entry happens to be evicted).

If you encode the scanner version number into the `key`, you'll only get a cache hit if the cached version matches the installed version. That's the behavior you want in combination with the later `if: steps.cache-sonar-scanner.outputs.cache-hit != 'true'`, since it'll install the updated version and update the cache when a different version number is present.

`restore-keys` is intended for use with partial matches; I don't think it's necessary to use it at all here if it's exactly equal to the `key.

If the scanner impl and its cache are going to be cached at the same cadence, you can also avoid some of the repetition by using `actions/cache@v2` instead of `@v1`; v2 supports passing multiple paths in the same step.

[GitHub's docs](https://docs.github.com/en/actions/guides/caching-dependencies-to-speed-up-workflows) are a good overview to learn more.

Reply:
Oh, I finally  get what you are saying.
Just pulled the caching as it saves almost no time.